rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,209 @@
+So, the synthesizer will help to verify the programs that require Kani to unwind loops for a very large number of times to cover all iterations.
+
+
+For example, the number of executed iterations of the loop in the following harness is dynamically bounded by an unbounded variable `y`.
+Only an unwinding value of `i32::MAX` can guarantee to cover all iteration of the loop, and hence satisfies the unwinding assertions.
+Unwinding the loop `i32::MAX` number of time will result in a too large goto program to be verified by CBMC.  
+```rust
+#[kani::proof]
+fn main() {
+    let mut y: i32 = kani::any_where(|i| *i > 0);
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+With the loop-contract synthesizer, Kani can synthesize the loop invariant `y >= 0`, with which it can prove the post-condition `y == 0` without unwinding the loop.
+
+
+Also, loop contracts could improve Kani’s verification time since all loops will be abstracted to a single iteration, as opposed to being unwound a large number of iterations.
+For example, we can easily find out that the following loop is bounded by an unwinding value of `5000`.
+Kani can verify the program in a few minutes by unwinding the loop 5000 times.
+With loop contracts, we only need to verify the single abstract iteration of the loop, which leads to a smaller query.
+As a result, Kani with the synthesizer can verify the program in a few seconds.
+```rust
+#[kani::proof]
+#[kani::unwind(5000)]
+fn main() {
+    let mut y: i32 = 5000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+The `goto-synthesizer` is an [enumeration-based synthesizer](https://www.cis.upenn.edu/~alur/SyGuS13.pdf).
+It enumerates candidate invariants from a pre-designed search space described by a given regular tree grammar and verifies if the candidate is an inductive invariant.
+Therefore it has the following limitations:
+1. the search space is not complete, so it may fail to find a working candidate. The current search space consists of only conjunction of linear inequalities built from the variables in the loop, which is not expressive enough to capture all loop invariants.
+For example, the loop invariant `a[i] == 0` contains array access and cannot be captured by the current search space.
+However, we can easily extend the search space to include more complex expressions with the cost of an exponential increase of the running time of the synthesizer.
+2. the synthesizer suffers from the same limitation as the loop contract verification in CBMC. For example, it does not support unbounded quantifiers, or dynamic allocations in the loop body. 
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to use the new command-line flag `--synthesize-loop-contracts` to run the synthesizer, which will attempt to synthesize loop contracts, and verify programs with the synthesized loop contracts.
```

Comment:
Why are we discussing stabilization now? In my opinion, this proposal should cover the unstable feature (enabled through `--Zsynthesize-loop-contracts` or similar?), and once it's ready to stabilize we can discuss further.